reconcile/vault_replication.py

```diff
@@ -0,0 +1,231 @@
+        deep_copy_versions(
+            dry_run=dry_run,
+            source_vault=source_vault,
+            dest_vault=dest_vault,
+            current_dest_version=0,
+            current_source_version=version,
+            path=path,
+        )
+
+
+def check_invalid_paths(
+    path_list: List[str],
+    policy_paths: Optional[List[str]],
+) -> None:
+
+    if policy_paths is not None:
+        invalid_paths = list_invalid_paths(path_list, policy_paths)
+        if invalid_paths:
+            logging.error(["replicate_vault_secret", "Invalid paths", invalid_paths])
+            raise VaultInvalidPaths
+
+
+def list_invalid_paths(path_list: List[str], policy_paths: List[str]) -> List[str]:
+    invalid_paths = []
+
+    for path in path_list:
+        if not policy_contains_path(path, policy_paths):
+            invalid_paths.append(path)
+
+    return invalid_paths
+
+
+def policy_contains_path(path: str, policy_paths: List[str]) -> bool:
+    return any(path in p_path for p_path in policy_paths)
+
+
+def get_policy_paths(policy_name, instance_name) -> List[str]:
+    query_data = vault_policies.query(query_func=gql.get_api().query)
+    policy_paths = []
+
+    if query_data.policy:
+        for policy in query_data.policy:
+            if policy.name == policy_name and policy.instance.name == instance_name:
+                for line in policy.rules.split("\n"):
+                    res = re.search(r"path \s*[\'\"](.+)[\'\"]", line)
+
+                    if res is not None:
+                        policy_paths.append(res.group(1))
+
+    return policy_paths
+
+
+def get_jenkins_secret_list(jenkins_instance: str) -> List[str]:
+    secret_list = []
+    query_data = jenkins_configs.query(query_func=gql.get_api().query)
+
+    if query_data.jenkins_configs:
+        for p in query_data.jenkins_configs:
+            if (
+                isinstance(p, JenkinsConfigV1_JenkinsConfigV1)
+                and p.instance.name == jenkins_instance
+                and p.config_path
+            ):
+                secret_paths = [
+                    line
+                    for line in p.config_path.content.split("\n")
+                    if "secret-path" in line
+                ]
+                for line in secret_paths:
+                    res = re.search(r"secret-path:\s*[\'\"](.+)[\'\"]", line)
+                    if res is not None:
+                        secret_list.append(res.group(1))
+
+    return secret_list
+
+
+def get_vault_credentials(vault_instance: VaultInstanceV1):
+    vault_creds = {"server": "", "role_id": None, "secret_id": None}
+    vault = cast(_VaultClient, VaultClient())
+
+    vault_instance_auth = cast(VaultInstanceAuthApproleV1, vault_instance.auth)
+
+    role_id = {
+        "path": vault_instance_auth.role_id.path,
+        "field": vault_instance_auth.role_id.field,
+    }
+    secret_id = {
+        "path": vault_instance_auth.secret_id.path,
+        "field": vault_instance_auth.secret_id.field,
+    }
```

Comment:
Can't you use directly `vault_instance.auth`?

Reply:
The reason lies here:

```gql
          auth {
            provider
            secretEngine
            ... on VaultInstanceAuthApprole_v1 {
              roleID {
                path
                field
              }
              secretID {
                path
                field
              }
            }
          }
```

By graphql standard, this inline query can return 2 kinds of auths: one is `VaultInstanceAuthApprole_v1` and the other is the interface itself (which is everything except `VaultInstanceAuthApprole_v1`). I.e., your auth object must be a Union to properly abstract this. The underlying problem with inline queries is, that they are not guaranteed to only return objects that are defined by the inline - they still return anything that is not defined by the inline.

My understanding is you could also make mypy happy by using `isinstance()` to filter any none `VaultInstanceAuthApprole_v1`

Reply:
d'oh, that lack of guarantee sucks. I mean, the whole point of using the `... on VaultInstanceAuthApprole_v1` is to only return objects of that kind 😂

Reply:
Another funny aspect of gql is, even if you pull in all attributes into the inlined fragment like this:

```gql
    auth {
            ... on VaultInstanceAuthApprole_v1 {
              provider
              secretEngine
              roleID {
                path
                field
              }
              secretID {
                path
                field
              }
            }
          }
```

You'd still require a Union with the Interface (which has no fields then), because the gql server will still return `{}` for any none `VaultInstanceAuthApprole_v1` object. I.e., once you go inline fragment query, you must deal with the Union.

Reply:
My understanding is that in this specific scenario we only have a single auth provider: approle. I verified by firing these 2 queries:

```gql
query VaultInstances {
  vault_instances: vault_instances_v1 {
    name
    address
    auth {
      provider
      secretEngine
      ... on VaultInstanceAuthApprole_v1 {
        roleID {
          path
          field
        }
      }
    }
  }
}
```

```gql
query VaultInstances {
  vault_instances: vault_instances_v1 {
    name
    address
    auth {
      provider
      secretEngine
    }
  }
}
```

Both return exactly the same objects, but with different fields.

I.e., in this specific case the Unions seems redundant indeed, because we only have approle providers. However, if in app-interface we are to add another auth provider, then the above **fully inlined** query would return `{}` or `{"provider": "sth", "name": "sth"}` for anything which is not an approle provider. That will lead to issues if there is no Union to catch it.

Reply:
hmm. the schema allows for two types: https://github.com/app-sre/qontract-schemas/blob/main/graphql-schemas/schema.yml#L2310-L2320

But we only have approle in app-interface. I don't know why we support both. Maybe fedramp 🤷

Reply:
So your recommendation is to use `isinstance` instead of casting, correct @fishi0x01 ?

Reply:
Yes, we talked about it on the call this morning, I'm doing the changes :)

Reply:
Yes `isinstance` will make mypy happy because it filters the desired objects in a safe way. `cast` will lead to undesired results, once we have different auth provider classes returned.

Reply:
Changed cast to `isinstance()` now we should be safe, thanks!